### DISCUSSION TITLE
How can I specify a custom global environment with Deno.emit?

### DISCUSSION THREAD

**Discussion by jsejcksn** [denoland/deno#10549] [Q&A]
I see that I can supply predefined environment enums in the emit options `compilerOptions.lib` array (e.g. `ES2017`/`DOM.Iterable`, etc.), but I don't see a way to provide other environment types (e.g. paths to declaration files) because properties like `compilerOptions.typeRoots` aren't supported.

If I want to compile for other [custom] runtime environments, do I really have to import the types into every source file or is there a configuration alternative?

**Comment by kitsonk** [comment]
Does something like this not work for you?

```ts
Deno.emit("test.ts", {
  compilerOptions: {
    types: ["./example/env.d.ts"]
  }
});
```

**Reply by jsejcksn** [reply]
@kitsonk I will try that option and report back. I didn't engage this option previously because I can't understand the description in the JSDoc:

> The type definitions are resolved according to the normal Deno resolution irrespective of if sources are provided on the call.

- Are the relative paths relative to CWD?
- Is using this option the equivalent of: automatically injecting an import statement for every string in the array at the top of each source file? e.g.:

```ts
import "./example/env.d.ts";
```

It's difficult to understand without a concrete example.

**Reply by jsejcksn** [reply]
It doesn't seem to work. I get the following kind of error for each path specified in `compilerOptions.types`:

```
TS2688 [ERROR]: Cannot find type definition file for './types/shared-node-browser.d.ts'.
  The file is in the program because:
    Entry point of type library './types/shared-node-browser.d.ts' specified in compilerOptions
```

**Reply by kitsonk** [reply]
> - Are the relative paths relative to CWD?

They follow the same resolution as if you passed it in `Deno.emit("./some_file.ts")`.

> - Is using this option the equivalent of: automatically injecting an import statement for every string in the array at the top of each source file?

No, not exactly.  [More like](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html):

```ts
/// <reference path="./example/env.d.ts" />
```

The `/// <reference types="">` directive works as well, but like the rest of Deno, no "magic" resolution, which is why `typeRoots` make no sense in a Deno context and the `types` triple-slash directive is actually redundant.

A concrete example would be specific to what you are trying to accomplish.

PRs to clarify the documentation or examples are welcome.

**Reply by jsejcksn** [reply]
@kitsonk See my message above. Here's an attached minimal example so you can reproduce. The `./run` script has all the necessary args to run the example.

[example.zip](https://github.com/denoland/deno/files/6448922/example.zip)

If I include

```ts
import "./types/mod.d.ts";
```

at the top of every source file, it compiles without errors, but when only using the `types` property as you suggested (and not manually including import statements), I get errors ☹️ :

```
example % ./run 
TS2688 [ERROR]: Cannot find type definition file for './types/shared-node-browser.d.ts'.
  The file is in the program because:
    Entry point of type library './types/shared-node-browser.d.ts' specified in compilerOptions

TS2584 [ERROR]: Cannot find name 'console'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
  console.log('hello world');
  ~~~~~~~
    at file:///Users/jesse/example/mod.ts:2:3

Found 2 errors.
"mod.bundle.cjs" written
```

Any ideas?

**Reply by jsejcksn** [reply]
@kitsonk Do you have any new insights about the scenario?
